core/modules/exact_match.cc

```diff
@@ -84,33 +84,23 @@ pb_error_t ExactMatch::Init(const bess::pb::ExactMatchArg &arg) {
   num_fields_ = arg.fields_size();
   total_key_size_ = align_ceil(size_acc, sizeof(uint64_t));
 
-  int ret = ht_.Init(total_key_size_, sizeof(gate_idx_t));
-  if (ret < 0) {
-    return pb_error(-ret, "hash table creation failed");
-  }
-
   return pb_errno(0);
 }
 
-void ExactMatch::DeInit() {
-  ht_.Close();
-}
-
 void ExactMatch::ProcessBatch(bess::PacketBatch *batch) {
   gate_idx_t default_gate;
   gate_idx_t out_gates[bess::PacketBatch::kMaxBurst];
-
-  int key_size = total_key_size_;
-  char keys[bess::PacketBatch::kMaxBurst][HASH_KEY_SIZE] __ymm_aligned;
+  em_hkey_t keys[bess::PacketBatch::kMaxBurst] __ymm_aligned;
 
   int cnt = batch->cnt();
 
-  default_gate = ACCESS_ONCE(default_gate_);
-
   for (int i = 0; i < cnt; i++) {
-    memset(&keys[i][key_size - 8], 0, sizeof(uint64_t));
+    char *p = reinterpret_cast<char *>(&keys[i]);
+    memset(p + total_key_size_ - 8, 0, sizeof(uint64_t));
   }
 
+  default_gate = ACCESS_ONCE(default_gate_);
```

Comment:
default_gate_ might be updated during ProcessBatch() by a gRPC thread (with .mt_safe=1 command). ACCESS_ONCE ensures that the variable is read only once at the beginning of ProcessBatch(), so that the value does not change in the middle (since it is stored in the local variable).

ACCESS_ONCE is a macro that temporarily makes a variable `volatile`; compiler won't try to optimize its access. Otherwise, default_gate_ can be read from memory multiple times, potentially with different values.